Adds Ruby v3.2.0

san983 opened the pull request:
Closes #114.

Also fixes ERB.new warnings in generate-ci-cd-yaml.rb like:

~~~
Regenerating .github/workflows/ci-cd-publish-test-production.yml
./internal-scripts/generate-ci-cd-yaml.rb:53: warning: Passing safe_level with the 2nd argument of ERB.new is deprecated. Do not use it, and specify other arguments as keyword arguments.
./internal-scripts/generate-ci-cd-yaml.rb:53: warning: Passing trim_mode with the 3rd argument of ERB.new is deprecated. Use keyword argument like ERB.new(str, trim_mode: ...) instead.
./internal-scripts/generate-ci-cd-yaml.rb:53: warning: Passing eoutvar with the 4th argument of ERB.new is deprecated. Use keyword argument like ERB.new(str, eoutvar: ...) instead.
~~~

FooBarWidget commented:
Thanks! Looks like the jemalloc patch needs to be updated for Ruby 3.2:

~~~
Configuring and compiling Ruby...
------------------------------------------
+ patch -p1 -ti /system/resources/ruby_31_jemalloc.patch
patching file version.c
Hunk #1 FAILED at 41.
1 out of 1 hunk FAILED -- saving rejects to file version.c.rej
~~~

Also I heard that YJIT is now production-ready but requires Rust. If so then we should add Rust into our build environment images.

FooBarWidget commented:
Also, it looks like the ERB.new deprecation fix is not backwards-compatible. On 3.1 I get:

~~~
irb(main):001:0> ERB.new('', safe_level: 0)
ruby/3.1.0/erb.rb:811:in `initialize': unknown keyword: :safe_level (ArgumentError)
~~~

I don't have a problem with bumping the required Ruby version to 3.2, but that needs to be codified/documented in the Github workflows (the Ruby version passed to the setup-ruby action) as well as dev-handbook/dev-environment-setup.md.

san983 commented:
> Thanks! Looks like the jemalloc patch needs to be updated for Ruby 3.2:
> 
> ```
> Configuring and compiling Ruby...
> ------------------------------------------
> + patch -p1 -ti /system/resources/ruby_31_jemalloc.patch
> patching file version.c
> Hunk #1 FAILED at 41.
> 1 out of 1 hunk FAILED -- saving rejects to file version.c.rej
> ```
> 
> Also I heard that YJIT is now production-ready but requires Rust. If so then we should add Rust into our build environment images.

Hi @FooBarWidgetm, just added:

- `rustc` library in build environments + build JYIT (as per https://github.com/ruby/ruby/blob/master/doc/yjit/yjit.md#building-yjit)
- Updated bundler to all Gemfiles
- Fixed a build flag for no docs

WDYT?

FooBarWidget commented on environments/centos-7/Dockerfile:
```diff
@@ -14,6 +14,7 @@ RUN set -x && \
         gcc gcc-c++ make patch bzip2 curl autoconf automake \
         openssl-devel libyaml-devel libffi-devel readline-devel zlib-devel \
         gdbm-devel ncurses-devel && \
+        rustc && \
```
No need to update centos-7. 3.0+ is excluded in config.yml anyway.

san983 commented on resources/ruby_32_jemalloc.patch:
```diff
@@ -0,0 +1,17 @@
++const char ruby_description[] = RUBY_DESCRIPTION_WITH(" +jemalloc");
++static const char ruby_description_with_mjit[] = RUBY_DESCRIPTION_WITH(" +jemalloc +MJIT");
++static const char ruby_description_with_yjit[] = RUBY_DESCRIPTION_WITH(" +jemalloc " YJIT_DESCRIPTION);
+ const char ruby_copyright[] = RUBY_COPYRIGHT;
```
Oops fixed at 3bdd7d7

san983 commented on environments/debian-11/Dockerfile:
```diff
@@ -14,6 +14,7 @@ RUN set -x && \
         dpkg-dev curl ca-certificates \
         libssl-dev libyaml-dev libreadline-dev zlib1g-dev \
         libncurses5-dev libffi-dev libgdbm6 libgdbm-dev && \
+        rustc && \
```
Done at 4769e8a

san983 commented on container-entrypoints/build-ruby:
```diff
@@ -86,23 +86,29 @@ if [[ "$VARIANT" = jemalloc ]]; then
+    elif [[ "$PACKAGE_VERSION" =~ ^3\.1 ]]; then
         run patch -p1 -ti /system/resources/ruby_31_malloctrim.patch
+    else
+        EXTRA_CONFIGURE_ARGS+=(--enable-yjit)
```
What about the ones I added at 13f22e4?

san983 reviewed (commented):
Thanks @FooBarWidget. I've just addressed your feedback.

FooBarWidget commented on environments/el-9/Dockerfile:
```diff
@@ -13,7 +13,8 @@ RUN set -x && \
         openssl-devel libyaml-devel libffi-devel readline-devel zlib-devel \
-        gdbm-devel ncurses-devel && \
+        gdbm-devel ncurses-devel \
+        rustc && \
```
Should be rust-toolchain.

san983 replied:
Fixed at 86ce853

FooBarWidget commented on environments/utility/Dockerfile:
```diff
@@ -29,7 +29,7 @@ RUN set -x && \
 
 RUN set -x && \
-    gem install bundler --no-document -v 2.2.28 && \
+    gem install bundler --no-document -v 2.4.1 && \
```
You need to bump the corresponding image_tag.

FooBarWidget commented on internal-scripts/test-debs:
```diff
@@ -280,6 +280,17 @@ if [[ "$VARIANT" = jemalloc ]]; then
     run c++ "-I$INCLUDE_DIR" -Wall /system/resources/test-env/jemalloc-cxx.cpp -o /tmp/jemalloc_cxx
 fi
 
+if [[ "$RUBY_PACKAGE_VERSION" =~ ^3\.2 ]]; then
```
Rather than checking whether version is 3.2.x, the check should be version >= 3.2.

san983 replied:
Added at 8ee5879

FooBarWidget commented on environments/centos-8/image_tag:
```diff
@@ -1,4 +1,4 @@
 # to force users to re-pull the image (e.g. when your change adds
 # a feature that our scripts rely on, or is breaking).
-1
+2
```
You need to rerun ./internal-scripts/generate-ci-cd-yaml.rb. For convenience you should setup the Git precommit hook locally: https://github.com/fullstaq-ruby/server-edition/blob/main/dev-handbook/dev-environment-setup.md#git-hooks

san983 replied:
Thanks! fixed at 86ce853

FooBarWidget commented on internal-scripts/test-debs:
```diff
@@ -280,6 +280,20 @@ if [[ "$VARIANT" = jemalloc ]]; then
     run c++ "-I$INCLUDE_DIR" -Wall /system/resources/test-env/jemalloc-cxx.cpp -o /tmp/jemalloc_cxx
 fi
 
+MAJOR_VERSION=`cut -d '.' -f 1 <<< "$RUBY_PACKAGE_VERSION"`
```
To keep shellcheck happy, you should use `$(...)` instead of backticks.

FooBarWidget replied:
Ditto for test-rpms.

san983 replied:
Nice catch, tweaked at d5a1653!

san983 replied:
Nice catch, tweaked at d5a1653!

FooBarWidget reviewed (changes requested):
Looking good, almost there.

FooBarWidget commented on environments/centos-8/Dockerfile:
```diff
@@ -13,7 +13,8 @@ RUN set -x && \
         openssl-devel libyaml-devel libffi-devel readline-devel zlib-devel \
-        gdbm-devel ncurses-devel && \
+        gdbm-devel ncurses-devel \
+        rust-toolchain && \
```
Sorry, I was wrong. It's apparently rust-toolset, not rust-toolchain.

san983 replied:
Fixed at f602878

san983 replied:
Fixed at f602878

FooBarWidget commented on environments/utility/Dockerfile:
```diff
@@ -29,7 +29,7 @@ RUN set -x && \
 
 RUN set -x && \
-    gem install bundler --no-document -v 2.2.28 && \
+    gem install bundler --no-document -v 2.4.1 && \
```
Looks like this doesn't work. Bundler 2.4 requires Ruby 2.6, but Ubuntu 18.04 has Ruby 2.5.

We could upgrade the utility image to Ubuntu 22.04, but then we'll also need to migrate from createrepo to createrepo_c, which is a bigger change. It's easiest to just stick with Bundler 2.2 for now.

san983 replied:
Makes sense, rollbacked it in 21dbe8d

pftg commented:
Hey, great to see YJIT!!! Thanks!

FooBarWidget commented on environments/el-9/Dockerfile:
```diff
@@ -13,7 +13,8 @@ RUN set -x && \
         openssl-devel libyaml-devel libffi-devel readline-devel zlib-devel \
-        gdbm-devel ncurses-devel && \
+        gdbm-devel ncurses-devel \
+        rust-toolchain && \
```
centos-8 and all other Docker images now build properly. But this here (for el-9) needs to be changed to rust-toolset too.

FooBarWidget commented on container-entrypoints/build-ruby:
```diff
@@ -86,23 +86,29 @@ if [[ "$VARIANT" = jemalloc ]]; then
+    elif [[ "$PACKAGE_VERSION" =~ ^3\.1 ]]; then
         run patch -p1 -ti /system/resources/ruby_31_malloctrim.patch
+    else
+        EXTRA_CONFIGURE_ARGS+=(--enable-yjit)
```
It looks like YJIT requires Rust 2021 edition support. YJIT fails to build on the following distributions:

 * debian-9 (rustc 1.41.1)
 * debian-10 (rustc 1.16.0)
 * debian-11 (rustc 1.48.0)

Surprisingly it works on Ubuntu 18.04 (younger than Debian 10) because it has rustc 1.61.0. I tested locally before on Ubuntu 18.04 and compiling YJIT works so I assumed it would also work on Debian 10 and maybe even 9, but apparently not. :(

This is the error:

~~~
building Rust YJIT (release mode)
error: argument for `--edition` must be one of: 2015|2018. (instead was `2021`)
~~~

We have a couple of choices here:

 * Disable YJIT on certain distributions. If we do this then we'll also have to modify the tests to test for YJIT only on distributions where it's not disabled.
 * Install an up-to-date enough Rust in the debian-9 through 11 images. (Hopefully there are binaries...)

What do you think?

FooBarWidget replied:
I've pushed a commit to your branch that implements the latter strategy. It looks like rustup-init indeed installs binaries. Could you let me know what you think?

san983 replied:
Sorry the delay but looks awesome @FooBarWidget! Thanks!

Glad to see you're able to sort the debian ones.

It looks we have a build passing now :) 🚀

FooBarWidget commented on resources/ruby_32_malloctrim.patch:
```diff
@@ -0,0 +1,19 @@
+diff --git a/version.c b/version.c
```
I've pushed a fix to your branch.

FooBarWidget commented:
With the latest few commits, the CI passes. We're almost there.

san983 commented on environments/el-9/Dockerfile:
```diff
@@ -13,7 +13,8 @@ RUN set -x && \
         openssl-devel libyaml-devel libffi-devel readline-devel zlib-devel \
-        gdbm-devel ncurses-devel && \
+        gdbm-devel ncurses-devel \
+        rust-toolchain && \
```
Cool, thanks!

FooBarWidget commented:
It's released now. Thanks again @san983

micke commented:
Thank you @san983 and @FooBarWidget!
